pkg/verify/signature.go

```diff
@@ -104,10 +116,96 @@ func VerifySignatureWithArtifactDigests(sigContent SignatureContent, verificatio
 	return verifyEnvelopeWithArtifactDigests(verifier, envelope, digests)
 }
 
-func getSignatureVerifier(verificationContent VerificationContent, tm root.TrustedMaterial) (signature.Verifier, error) {
+// compatVerifier is a signature.Verifier that tries multiple verifiers
+// and returns nil if any of them verify the signature. This is used to
+// verify signatures that were generated with old clients that used SHA256
+// for ECDSA P384/P521 keys.
+type compatVerifier struct {
+	verifiers []signature.Verifier
+}
+
+func (v *compatVerifier) VerifySignature(signature, message io.Reader, opts ...signature.VerifyOption) error {
+	// Create a buffer to store the signature bytes
+	sigBuf := &bytes.Buffer{}
+	sigTee := io.TeeReader(signature, sigBuf)
+	sigBytes, err := io.ReadAll(sigTee)
+	if err != nil {
+		return fmt.Errorf("failed to read signature: %w", err)
+	}
+
+	// Create a buffer to store the message bytes
+	msgBuf := &bytes.Buffer{}
+	msgTee := io.TeeReader(message, msgBuf)
+	msgBytes, err := io.ReadAll(msgTee)
+	if err != nil {
+		return fmt.Errorf("failed to read message: %w", err)
+	}
+
+	for idx, verifier := range v.verifiers {
+		if idx != 0 {
+			fmt.Fprint(os.Stderr, "Failed to verify signature with default verifier, trying compatibility verifier\n")
+		}
+		err := verifier.VerifySignature(bytes.NewReader(sigBytes), bytes.NewReader(msgBytes), opts...)
+		if err == nil {
+			return nil
+		}
+	}
+	return fmt.Errorf("no compatible verifier found")
+}
+
+func (v *compatVerifier) PublicKey(opts ...signature.PublicKeyOption) (crypto.PublicKey, error) {
+	return v.verifiers[0].PublicKey(opts...)
+}
+
+func compatSignatureVerifier(leafCert *x509.Certificate, enableCompat bool) (signature.Verifier, error) {
+	// LoadDefaultSigner/Verifier functions accept a few options to select
+	// the default signer/verifier when there are ambiguities, like for
+	// ED25519 keys, which could be used with PureEd25519 or Ed25519ph.
+	//
+	// Pass `WithED25519ph()` to select Ed25519ph by default, when ED25519
+	// key is found, because for hashedrekord entries this is the only option.
+	defaultOpts := []signature.LoadOption{options.WithED25519ph()}
```

Comment:
This would mean this fallback verifier is only valid for hashedrekord and not DSSE, yes? We should document if so

Reply:
to be honest there's no runtime check or anything in place to enforce that this is done only for hashedrekord. It will just use pre-hashed every time there is a ed25519 key no matter what. Shall we restrict the behaviour to just hashedrekord?

Reply:
Can we pass the type in and set whether it’s the PH variant or not based on the type? There could also be DSSE bundles with the older mapping that we should support.

Reply:
Waiting for tests, but this should be addressed with the last commit.

Reply:
This looks great!